File: src/kontur/providers/python/ukrainian-tts.py
#!/usr/bin/env python3
"""
Ukrainian TTS Wrapper for ATLAS
Uses robinhad/ukrainian-tts package
"""

import sys
import argparse
import io
import logging

# Configure logging to stderr so stdout remains clean for audio data
logging.basicConfig(level=logging.INFO, format='%(name)s - %(levelname)s - %(message)s', stream=sys.stderr)
logger = logging.getLogger('ukrainian-tts')

try:
    from ukrainian_tts.tts import TTS
except ImportError:
    logger.error("ukrainian-tts package not found. Please install it with: pip install ukrainian-tts")
    sys.exit(1)

def main():
    parser = argparse.ArgumentParser(description='Ukrainian TTS Wrapper')
    parser.add_argument('--text', type=str, help='Text to speak. If not provided, reads from stdin.')
    parser.add_argument('--voice', type=str, default='tetiana', choices=['tetiana', 'mykyta', 'lada', 'dmytro', 'oleksa'], help='Voice to use (lowercase)')
    parser.add_argument('--output', type=str, help='Output file path. If not provided, writes wav to stdout.')

    args = parser.parse_args()

    # Get text
    text = args.text
    if not text:
        if not sys.stdin.isatty():
             text = sys.stdin.read().strip()
    
    if not text:
        logger.error("No text provided.")
        sys.exit(1)

    logger.info(f"Generating audio for: {text[:50]}... (Voice: {args.voice})")

    try:
        tts = TTS(device="cpu")
        
        if args.output:
            with open(args.output, mode="wb") as f:
                tts.tts(text, args.voice.lower(), 'dictionary', f)
            logger.info(f"Audio saved to {args.output}")
        else:
            buffer = io.BytesIO()
            tts.tts(text, args.voice.lower(), 'dictionary', buffer)
            buffer.seek(0)
            sys.stdout.buffer.write(buffer.read())
            sys.stdout.buffer.flush()
            
    except Exception as e:
        logger.error(f"TTS Generation failed: {e}")
        sys.exit(1)

if __name__ == '__main__':
    main()

File: src/kontur/providers/ukrainian-tts.ts

import { ITTSProvider, TTSRequest, TTSResponse, ProviderName } from './types';
import { spawn } from 'child_process';
import path from 'path';

export class UkrainianTTSProvider implements ITTSProvider {
    name: ProviderName = 'ukrainian';

    isAvailable(): boolean {
        // Ukrainian TTS is available if python3 is in the path
        // We assume it's available and let errors surface at runtime
        return true;
    }

    async speak(request: TTSRequest): Promise<TTSResponse> {
        return new Promise((resolve, reject) => {
            const scriptPath = path.join(__dirname, 'python', 'ukrainian-tts.py');

            // Get voice: from request, or per-agent env config, or default
            // Voice can be passed in request.voice as "tetiana" or agent name triggers lookup
            let voice = request.voice || 'tetiana';

            // If voice looks like an agent name, look up their configured voice
            const agentVoices: Record<string, string> = {
                'ATLAS': process.env.UKRAINIAN_VOICE_ATLAS || 'dmytro',
                'TETYANA': process.env.UKRAINIAN_VOICE_TETYANA || 'tetiana',
                'GRISHA': process.env.UKRAINIAN_VOICE_GRISHA || 'oleksa'
            };
            if (agentVoices[voice.toUpperCase()]) {
                voice = agentVoices[voice.toUpperCase()];
            }

            const pythonProcess = spawn('python3', [
                scriptPath,
                '--voice', voice
            ]);

            const audioChunks: Buffer[] = [];
            let errorData = '';

            // Send text to stdin
            pythonProcess.stdin.write(request.text);
            pythonProcess.stdin.end();

            pythonProcess.stdout.on('data', (chunk) => {
                audioChunks.push(chunk);
            });

            pythonProcess.stderr.on('data', (data) => {
                errorData += data.toString();
            });

            pythonProcess.on('close', (code) => {
                if (code !== 0) {
                    console.error('[UkrainianTTS] Python script error:', errorData);
                    reject(new Error(`Ukrainian TTS failed: ${errorData}`));
                } else {
                    const audioBuffer = Buffer.concat(audioChunks);
                    // Convert Node Buffer to ArrayBuffer for compatibility
                    const arrayBuffer = audioBuffer.buffer.slice(
                        audioBuffer.byteOffset,
                        audioBuffer.byteOffset + audioBuffer.byteLength
                    );

                    resolve({
                        audio: arrayBuffer,
                        mimeType: 'audio/wav',
                        provider: 'ukrainian'
                    });
                }
            });

            pythonProcess.on('error', (err) => {
                reject(new Error(`Failed to spawn python process: ${err.message}. Make sure python3 is installed and ukrainian-tts package is installed.`));
            });
        });
    }
}
